### ISSUE TITLE
Stack overflow when handling redefinition of complex type

### ISSUE THREAD

**Issue by matachi** [mganss/XmlSchemaClassGenerator#52]
Hi,

I'm trying to generate code from the following example. `Schema1.xsd`:

```xml
<?xml version="1.0"?>
<xs:schema xmlns:xs="http://www.w3.org/2001/XMLSchema">

<xs:complexType name="pname">
  <xs:sequence>
    <xs:element name="firstname"/>
    <xs:element name="lastname"/>
  </xs:sequence>
</xs:complexType>

<xs:element name="customer" type="pname"/>

</xs:schema>
```

`Schema2.xsd`:

```xml
<?xml version="1.0"?>
<xs:schema xmlns:xs="http://www.w3.org/2001/XMLSchema">

<xs:redefine schemaLocation="Schema1.xsd">
  <xs:complexType name="pname">
    <xs:complexContent>
      <xs:extension base="pname">
        <xs:sequence>
          <xs:element name="country"/>
        </xs:sequence>
      </xs:extension>
    </xs:complexContent>
  </xs:complexType>
</xs:redefine>

<xs:element name="author" type="pname"/>

</xs:schema>
```

My generator program looks as follows:

```c#
using System.Collections.Generic;
using XmlSchemaClassGenerator;

namespace ClassGenerator
{
    internal class Program
    {
        internal static void Main(string[] args)
        {
            var generator = new Generator
            {
                OutputFolder = "",
                NamespacePrefix = "Output",
            };
            generator.Generate(new List<string>(new[] {"Schema2.xsd"}));
        }
    }
}
```

However, when running it, it throws the error:

> An unhandled exception of type 'System.StackOverflowException' occurred in Unknown Module. occurred

I have managed to locate that it occurs in the class `TypeModel` in the method `GetAllDerivedTypes`, https://github.com/mganss/XmlSchemaClassGenerator/blob/master/XmlSchemaClassGenerator/TypeModel.cs#L401. It gets stuck there in an infinite recursive loop.

Would it be possible/easy to implement support for .xsd files that redefine complex types?

**Comment by mganss** [comment]
I think this is related to #50. The infinite recursion can be prevented easily but then this gets generated:
```C#
public partial class Pname : Pname
{
   [System.Xml.Serialization.XmlElementAttribute("country", Form=System.Xml.Schema.XmlSchemaForm.Unqualified)]
   public object Country { get; set; }
}
```
If #50 got fixed, then the two `Pname` would get unique names and then this could work.

**Comment by mganss** [comment]
Fixed in acb5719
